pennylane/io/qasm_interpreter.py

```diff
@@ -0,0 +1,346 @@
+            for func in call_stack:
+                # if there is a control in the stack
+                if (
+                    call_stack[-1].__class__.__name__ == "partial"
+                    and "control" in call_stack[-1].keywords
+                ):
+                    # if we are processing the control now
+                    if "control" in func.keywords:
+                        res.keywords["wires"] = [res.keywords["wires"][-1]]
+                    # i.e. qml.ctrl(qml.RX, (1))(2, wires=0)
+                    res = func(res.func)(**res.keywords) if res is not None else func
+                else:
+                    # i.e. qml.pow(qml.RX(1.5, wires=0), z=4)
+                    res = func(res) if res is not None else func()
+
+        return call
+
+    @staticmethod
+    def _require_wires(context):
+        """
+        Simple helper that checks if we have wires in the current context.
+
+        Args:
+            context (dict): The current context.
+
+        Raises:
+            NameError: If the context is missing a wire.
+        """
+        if "wires" not in context:
+            raise NameError(
+                f"Attempt to reference wires that have not been declared in {context['name']}"
+            )
+
+    def gate(self, gates_dict: dict, node: QASMNode, context: dict):
+        """
+        Registers a gate application. Builds a Callable partial
+        that can be executed when the QNode is called. The gate will be executed at that time
+        with the appropriate arguments.
+        TODO: a robust method for retrieving vars from context will be provided in follow-up PR according to [sc-90383]
+
+        Args:
+            node (QASMNode): The QuantumGate QASMNode.
+            context (dict): The current context.
+
+        Returns:
+            Callable: The Callable partial that will execute the gate with the appropriate arguments at
+                "runtime".
+
+        Raises:
+            NameError: If an argument is not found in the current context.
+        """
+        self._require_wires(context)
+        args = []
+        for arg in node.arguments:
+            if hasattr(arg, "name") and "vars" in context and arg.name in context["vars"]:
+                # the context at this point should reflect the states of the
+                # variables as evaluated in the correct (current) scope.
+                if context["vars"][arg.name]["val"] is not None:
+                    args.append(context["vars"][arg.name]["val"])
+                else:
+                    raise NameError(f"Attempt to reference uninitialized parameter {arg.name}!")
+            elif re.search("Literal", arg.__class__.__name__) is not None:
+                args.append(arg.value)
+            else:
+                raise NameError(
+                    f"Uninitialized variable {arg.name if hasattr(arg, 'name') else arg.__class__.__name__} "
+                    f"encountered in QASM."
+                )
```

Comment:
Can we create a separate method for retrieving the value of something? This looks very similar to the `PlxprInterpreter.read`, maybe you can take a look.
https://github.com/PennyLaneAI/pennylane/blob/ab5f118b0f1edeacc63fb46eb1e49cf9fba913c9/pennylane/capture/base_interpreter.py#L261

Reply:
That's fine. It looks like that PR adds a lot more to `retrieve_variable`. I think it would make both PRs more readable if you implement and use a skeleton of that in this PR, and that second PR would simply be adding to the method.